Updating Conda section (#59)
ARCTraining/hpc2-software: book/course/conda.md

Conda makes it easy to switch between different environments and allows you to create and delete them as required.
Conda environments also make it easier to share our environment setup between machines and with collaborators as we can export our environments into a text file.

> If you want to find out more about [good dependency management practises in general, please read our documentation](https://arcdocs.leeds.ac.uk/aire/usage/dependency_management.html#dependency-management); we use this material to inform this session but take a more trial-and-error approach here.

Reviewer: ```suggestion
If you want to find out more about [good dependency management practices in general, please read our documentation](https://arcdocs.leeds.ac.uk/aire/usage/dependency_management.html#dependency-management); we use this material to inform this session but take a more trial-and-error approach here.
```
I could be wrong though!

- **Export your exact environment as metadata for analysis results**: it is useful to save a snapshot of your environment to store along any results or outputs produced in that specific environment. 
- **Environments must be stored in your `home` directory and all research output must be stored in `/mnt/scratch/users`**: misuse of the system can affect performance for **all users** and will lead to your jobs being stopped.

(creating-environments)=
### Creating environments

There are two main ways to create a fresh Conda environment:
  1. Creating directly from the command line with a list of required packages;
  2. Creating from an `environment.yml` file that lists required packages.

We will step through examples of both, and compare both techniques.

#### 1. On the fly creation

If you have come across Conda before, this is likely the method of creating environments that you've encountered.

You can create an environment with Conda with the subcommand `conda create`.

> When creating an environment we need to give it a name; we recommend giving it a name related to the project you're going building it to support. In this example, we use the (unimaginative name) `py39-env` as we're going to be using Python 3.9; you can imagine that if you're working with multiple different versions of Python is could be useful to record this in the environment name, and prefix it with the project title.

Reviewer: ```suggestion
When creating an environment we need to give it a name; we recommend giving it a name related to the project you're building it to support. In this example, we use the (unimaginative name) `py39-env` as we're going to be using Python 3.9; you can imagine that if you're working with multiple different versions of Python is could be useful to record this in the environment name, and prefix it with the project title.
```

```
````

(installing-packages)=
### Updating a Conda environment and installing new packages

It's very likely that after creating an environment with a certain list of packages, you'll want to add other packages, or potentially change what version of a package you have installed.

Earlier we created the `data-sci-env` and installed some useful data science packages.
We've discovered we also need the `statsmodels` package for some extra work we want to do so we'll look at how to install this package within our existing environment.

````{admonition} Searching for packages

> Conda has a command-line search functionality that we describe below in the section [Use Conda to search for a package](#); you can also use the [`conda-forge` repository](https://anaconda.org/conda-forge) or [`bioconda` repository](https://anaconda.org/bioconda) to search for packages.

Reviewer: Does that (#) do what you want?

Author: ```suggestion
Conda has a command-line search functionality that we describe below in the section [Use Conda to search for a package](#searching-for-packages); you can also use the [`conda-forge` repository](https://anaconda.org/conda-forge) or [`bioconda` repository](https://anaconda.org/bioconda) to search for packages.
```

No, indeed it didn't - have added the link above

  statsmodels        conda-forge/linux-64::statsmodels-0.13.2-py310hde88566_0

Proceed ([y]/n)?
```
````

Conda will always prompt the user if we're happy to proceed with the installation and specifies all the other packages that will be installed or updated that are required for our specified package.
We confirm we wish to proceed by entering `y` and pressing Return.

````{admonition} View full output
:class: dropdown
```
Proceed ([y]/n)? y

Downloading and Extracting Packages
pytz-2022.2.1        | 224 KB    | ##################################### | 100%
libopenblas-0.3.21   | 10.1 MB   | ##################################### | 100%
scipy-1.9.1          | 26.2 MB   | ##################################### | 100%
patsy-0.5.2          | 188 KB    | ##################################### | 100%
statsmodels-0.13.2   | 11.2 MB   | ##################################### | 100%
pandas-1.4.4         | 12.5 MB   | ##################################### | 100%
numpy-1.23.2         | 7.1 MB    | ##################################### | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
```
````

> This installs any packages that are currently not installed (Conda caches packages locally incase they are required by other packages, this speeds up installs but uses more disk space to maintain this cache).

Reviewer: ```suggestion
This installs any packages that are currently not installed (Conda caches packages locally in case they are required by other packages, this speeds up installs but uses more disk space to maintain this cache).
```
